psycopg_c/psycopg_c/pq/pgconn.pyx

```diff
@@ -502,73 +604,104 @@ cdef class PGconn:
 
     def send_close_portal(self, const char *name) -> None:
         _check_supported("PQsendClosePortal", 170000)
-        _ensure_pgconn(self)
-        cdef int rv = libpq.PQsendClosePortal(self._pgconn_ptr, name)
+        cdef int rv
+        cdef libpq.PGconn *pgconn_ptr
+        with cython.critical_section(self):
+            pgconn_ptr = self._pgconn_ptr
+            if pgconn_ptr is not NULL:
+                rv = libpq.PQsendClosePortal(pgconn_ptr, name)
+        if pgconn_ptr is NULL:
+            raise e.OperationalError("the connection is closed")
         if not rv:
             raise e.OperationalError(
                 f"sending close prepared failed: {self.get_error_message()}"
             )
 
     def get_result(self) -> "PGresult" | None:
-        cdef libpq.PGresult *pgresult = libpq.PQgetResult(self._pgconn_ptr)
+        cdef libpq.PGresult *pgresult
+        with cython.critical_section(self):
+            pgresult = libpq.PQgetResult(self._pgconn_ptr)
```

Comment:
In a few function you are not using the local pointer copy. Yes, it seems that in those functions we were not using `_ensure_pgconn()` previously, but I wonder if we shouldn't try to use a consistent pattern.

Reply:
Not sure whether we should be doing additional work unless it's necessary. Are there specific APIs where a NULL connection pointer is okay while, in others, it's not?